File: fluentcore/src/main/java/com/mcreater/fxfluent/brush/AbstractColorBrush.java
package com.mcreater.fxfluent.brush;

import com.mcreater.fxfluent.xaml.ResourceDict;
import com.mcreater.fxfluent.xaml.XAMLManager;
import javafx.animation.Interpolatable;
import javafx.scene.layout.Region;
import javafx.scene.paint.CycleMethod;
import javafx.scene.paint.LinearGradient;
import javafx.scene.paint.Paint;
import javafx.scene.paint.Stop;

import java.util.List;
import java.util.Vector;
import java.util.function.BiConsumer;

public abstract class AbstractColorBrush implements Interpolatable<AbstractColorBrush>, BiConsumer<Region, BiConsumer<Region, AbstractColorBrush>> {
    public AbstractColorBrush interpolate(AbstractColorBrush abstractColorBrush, double v) {
        if (v == 0) return this;
        if (v == 1) return abstractColorBrush;
        if (abstractColorBrush instanceof SolidColorBrush && this instanceof SolidColorBrush) {
            return new SolidColorBrush(toSolidColorBrush().getPaint().interpolate(abstractColorBrush.toSolidColorBrush().getPaint(), v));
        }
        if (abstractColorBrush instanceof LinearGradientColorBrush) {
            LinearGradient end = abstractColorBrush.toLinearGradientColorBrush().getPaint();
            LinearGradient start = this.toLinearGradientColorBrush().getPaint();
            List<Stop> stops = new Vector<>();
            end.getStops().forEach(stop -> {
                final double[] offsetdelta = {Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY};
                final Stop[] result = {stop};
                start.getStops().forEach(s -> {
                    double noff = Math.abs(stop.getOffset() - s.getOffset());
                    if (noff < offsetdelta[0]) {
                        offsetdelta[0] = noff;
                        offsetdelta[1] = stop.getOffset() - s.getOffset();
                        result[0] = s;
                    }
                });
                stops.add(new Stop(
                        result[0].getOffset() + (stop.getOffset() - result[0].getOffset()) * v,
                        result[0].getColor().interpolate(stop.getColor(), v)
                ));
            });
            double startX = start.getStartX() + (end.getStartX() - start.getStartX()) * v;
            double startY = start.getStartY() + (end.getStartY() - start.getStartY()) * v;
            double endX = start.getEndX() + (end.getEndX() - start.getEndX()) * v;
            double endY = start.getEndY() + (end.getEndY() - start.getEndY()) * v;

            return new LinearGradientColorBrush(
                    new LinearGradient(
                            startX, startY, endX, endY, true, CycleMethod.NO_CYCLE,
                            stops
                    )
            );
        }
        return abstractColorBrush;
    }

    private SolidColorBrush toSolidColorBrush() {
        return (SolidColorBrush) this;
    }
    private LinearGradientColorBrush toLinearGradientColorBrush() {
        try {
            return (LinearGradientColorBrush) this;
        }
        catch (Exception ignored) {}

        return new LinearGradientColorBrush(
                new LinearGradient(
                        0, 0, 0, 3, true, CycleMethod.NO_CYCLE,
                        new Stop(0, toSolidColorBrush().getPaint()), new Stop(1, toSolidColorBrush().getPaint())
                )
        );
    }

    public abstract Paint getPaint();
    public void accept(Region region, BiConsumer<Region, AbstractColorBrush> regionAbstractColorBrushBiConsumer) {
        regionAbstractColorBrushBiConsumer.accept(region, this);
    }

    public static AbstractColorBrush find(String k) {
        return XAMLManager.getCurrentDict().findColorBrush(k);
    }

    public static AbstractColorBrush find(String k, ResourceDict dict) {
        return dict.findColorBrush(k);
    }
}

File: fluentcore/src/main/java/com/mcreater/fxfluent/brush/AbstractColorBrush.kt
package com.mcreater.fxfluent.brush

import javafx.animation.Interpolatable
import javafx.scene.layout.Region
import javafx.scene.paint.CycleMethod
import javafx.scene.paint.LinearGradient
import javafx.scene.paint.Paint
import javafx.scene.paint.Stop
import java.util.*
import java.util.function.BiConsumer
import java.util.function.Consumer
import kotlin.math.abs


abstract class AbstractColorBrush : Interpolatable<AbstractColorBrush?>,
    BiConsumer<Region?, BiConsumer<Region?, AbstractColorBrush?>> {
    override fun interpolate(abstractColorBrush: AbstractColorBrush?, v: Double): AbstractColorBrush? {
        if (v == 0.0) return this
        if (v == 1.0) return abstractColorBrush
        if (abstractColorBrush is SolidColorBrush && this is SolidColorBrush) {
            return SolidColorBrush(
                toSolidColorBrush().getPaint()!!.interpolate(
                    abstractColorBrush.getPaint(),
                    v
                )
            )
        }
        if (abstractColorBrush is LinearGradientColorBrush) {
            val end: LinearGradient = abstractColorBrush.getPaint()!!
            val start: LinearGradient = toLinearGradientColorBrush().getPaint()!!
            val stops: MutableList<Stop> = Vector()
            end.stops.forEach(Consumer { stop: Stop ->
                val offsetdelta = doubleArrayOf(
                    Double.POSITIVE_INFINITY,
                    Double.POSITIVE_INFINITY
                )
                val result = arrayOf(stop)
                start.stops
                    .forEach(Consumer { s: Stop ->
                        val noff = abs(stop.offset - s.offset)
                        if (noff < offsetdelta[0]) {
                            offsetdelta[0] = noff
                            offsetdelta[1] = stop.offset - s.offset
                            result[0] = s
                        }
                    })
                stops.add(
                    Stop(
                        result[0].offset + (stop.offset - result[0].offset) * v,
                        result[0].color.interpolate(stop.color, v)
                    )
                )
            })
            val startX = start.startX + (end.startX - start.startX) * v
            val startY = start.startY + (end.startY - start.startY) * v
            val endX = start.endX + (end.endX - start.endX) * v
            val endY = start.endY + (end.endY - start.endY) * v
            return LinearGradientColorBrush(
                LinearGradient(
                    startX, startY, endX, endY, true, CycleMethod.NO_CYCLE,
                    stops
                )
            )
        }
        return abstractColorBrush
    }

    private fun toSolidColorBrush(): SolidColorBrush {
        return this as SolidColorBrush
    }

    private fun toLinearGradientColorBrush(): LinearGradientColorBrush {
        try {
            return this as LinearGradientColorBrush
        } catch (ignored: Exception) {
        }
        return LinearGradientColorBrush(
            LinearGradient(
                0.0, 0.0, 0.0, 3.0, true, CycleMethod.NO_CYCLE,
                Stop(0.0, toSolidColorBrush().getPaint()), Stop(1.0, toSolidColorBrush().getPaint())
            )
        )
    }

    abstract fun getPaint(): Paint?

    override fun accept(region: Region?, regionAbstractColorBrushBiConsumer: BiConsumer<Region?, AbstractColorBrush?>) {
        regionAbstractColorBrushBiConsumer.accept(region, this)
    }
}

